### PR TITLE
feat: Generate a fallback report upon recursion limit hit

### PR DESCRIPTION
**Pull request by xunliu**
## Pull request overview
This PR implements a graceful fallback mechanism for handling recursion limit errors in agent execution. When an agent hits the LangGraph recursion limit, instead of failing completely, the system now generates a summary report using the partial results accumulated during execution.

### Changes:

- Adds _handle_recursion_limit_fallback() function to generate fallback reports from accumulated agent messages
- Modifies _execute_agent_step() to use streaming instead of ainvoke to capture partial results before recursion limit is hit
- Introduces enable_recursion_fallback configuration flag (default: True) to control the fallback behavior
- Refactors apply_prompt_template() by extracting get_system_prompt_template() helper function
- Adds new recursion_fallback.md prompt template for generating fallback summaries

### CODE REVIEW

File: src/graph/nodes.py [hunk 1]
```diff
@@ -1087,9 +1202,11 @@ async def _execute_agent_step(
             },
             goto="research_team",
         )
+    else:
+        response_messages = result["messages"]
```
[the remark this answers is not in the document]
Comment (xunliu) [reply src/graph/nodes.py]: fixed

File: src/prompts/template.py [hunk 2]
```diff
@@ -61,6 +60,16 @@ def apply_prompt_template(
...
+    try:
+        system_prompt = get_system_prompt_template(prompt_name, state, configurable, locale)
+        return [{"role": "system", "content": system_prompt}] + state["messages"]
+    except Exception as e:
+        raise ValueError(f"Error applying template {prompt_name} for locale {locale}: {e}")
+
+def get_system_prompt_template(
+    prompt_name: str, state: AgentState, configurable: Configuration = None, locale: str = "en-US"
+) -> str:
+
```
[the remark this answers is not in the document]
Comment (xunliu) [reply src/prompts/template.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 3]
```diff
@@ -0,0 +1,635 @@
...
+        partial_agent_messages = [HumanMessage(content="Test")]
+
+        # Mock unsanitized response with extra tokens
+        mock_llm_response = MagicMock()
+        mock_llm_response.content = "<extra_tokens>Summary content</extra_tokens>"
+
+        sanitized_content = "Summary content"
+
+        with patch("src.graph.nodes.get_llm_by_type") as mock_get_llm, \
+             patch("src.graph.nodes.get_system_prompt_template", return_value=[]), \
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+    async def test_fallback_raises_on_llm_failure(self):
+        """Test that fallback raises exception when LLM call fails."""
+        state = State(messages=[], locale="en-US")
+        current_step = MagicMock()
+
+        # Create test messages (not empty)
+        partial_agent_messages = [HumanMessage(content="Test")]
+
+        with patch("src.graph.nodes.get_llm_by_type") as mock_get_llm, \
+             patch("src.graph.nodes.get_system_prompt_template", return_value=[]):
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 5]
```diff
@@ -0,0 +1,635 @@
...
+        state = State(messages=[], locale="en-US")
+
+        # Create test messages (not empty)
+        partial_agent_messages = [HumanMessage(content="Test")]
+
+        mock_llm_response = MagicMock()
+        mock_llm_response.content = "Agent summary"
+
+        with patch("src.graph.nodes.get_llm_by_type") as mock_get_llm, \
+             patch("src.graph.nodes.get_system_prompt_template", return_value=[]), \
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 6]
```diff
@@ -0,0 +1,635 @@
...
+        """Test that fallback handles empty partial_agent_messages."""
+        state = State(messages=[], locale="en-US")
+        current_step = MagicMock()
+        partial_agent_messages = []  # Empty
+
+        mock_llm_response = MagicMock()
+        mock_llm_response.content = "Fallback summary"
+
+        with patch("src.graph.nodes.get_llm_by_type") as mock_get_llm, \
+             patch("src.graph.nodes.get_system_prompt_template", return_value=[]), \
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+        state = State(messages=[], locale="en-US")
+        current_step = MagicMock()
+        # Create non-empty messages to avoid early return
+        partial_agent_messages = [HumanMessage(content="Test")]
+
+        mock_llm_response = MagicMock()
+        mock_llm_response.content = "Summary"
+
+        with patch("src.graph.nodes.get_llm_by_type") as mock_get_llm, \
+             patch("src.graph.nodes.get_system_prompt_template", return_value=[]), \
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+        state = State(messages=[], locale="en-US")
+        current_step = MagicMock()
+        agent_input = {"messages": []}
+        partial_agent_messages = []
+
+        mock_llm_response = MagicMock()
+        mock_llm_response.content = "No observations available"
+
+        with patch("src.graph.nodes.get_llm_by_type") as mock_get_llm, \
+             patch("src.graph.nodes.get_system_prompt_template", return_value=[]), \
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: src/prompts/template.py [hunk 9]
```diff
@@ -74,15 +83,15 @@ def apply_prompt_template(
...
             # Fallback to English template
             template = env.get_template(f"{prompt_name}.md")
-        
+
         system_prompt = template.render(**state_vars)
-        return [{"role": "system", "content": system_prompt}] + state["messages"]
+        return system_prompt
     except Exception as e:
-        raise ValueError(f"Error applying template {prompt_name} for locale {locale}: {e}")
+        raise ValueError(f"Error get template {prompt_name} for locale {locale}: {e}")
```
[the remark this answers is not in the document]
Comment (xunliu) [reply src/prompts/template.py]: fixed

File: src/graph/nodes.py [hunk 10]
```diff
@@ -929,6 +930,80 @@ def validate_web_search_usage(messages: list, agent_name: str = "agent") -> bool
...
+    """
+    logger.warning(
+        f"Recursion limit reached for {agent_name} agent. "
+        f"Attempting graceful fallback with {len(messages)} accumulated messages."
+    )
+
+    if len(messages) == 0:
+        return messages
+
+
```
[the remark this answers is not in the document]
Comment (xunliu) [reply src/graph/nodes.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 11]
```diff
@@ -0,0 +1,635 @@
...
+            HumanMessage(content="Input message"),
+            AIMessage(content="", tool_calls=[tool_call]),
+            HumanMessage(content="Tool result: Search completed"),
+        ]
+
+        mock_llm_response = MagicMock()
+        mock_llm_response.content = "Fallback summary"
+
+        with patch("src.graph.nodes.get_llm_by_type") as mock_get_llm, \
+             patch("src.graph.nodes.get_system_prompt_template", return_value=[]), \
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 12]
```diff
@@ -0,0 +1,635 @@
...
+        state = State(messages=[], locale="en-US")
+        current_step = MagicMock()
+        agent_input = {"messages": []}
+        partial_agent_messages = []
+
+        mock_llm_response = MagicMock()
+        mock_llm_response.content = "Summary"
+
+        with patch("src.graph.nodes.get_llm_by_type") as mock_get_llm, \
+             patch("src.graph.nodes.get_system_prompt_template", return_value=[]), \
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 13]
```diff
@@ -0,0 +1,635 @@
...
+        with patch("src.graph.nodes.get_llm_by_type") as mock_get_llm, \
+             patch("src.graph.nodes.get_system_prompt_template") as mock_get_system_prompt, \
+             patch("src.graph.nodes.sanitize_tool_response", return_value=mock_llm_response.content):
+
+            mock_llm = MagicMock()
+            mock_llm.invoke = MagicMock(return_value=mock_llm_response)
+            mock_get_llm.return_value = mock_llm
+            mock_get_system_prompt.return_value = "Template"
+
+            result = await _handle_recursion_limit_fallback(
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 14]
```diff
@@ -0,0 +1,635 @@
...
+        current_step = MagicMock()
+        current_step.execution_res = None
+
+        # Mock agent input with tool observations
+        agent_input = {
+            "messages": [
+                HumanMessage(content="# Research Topic\n\nAI safety\n\n# Current Step\n\n## Title\n\nAnalyze AI safety"),
+            ]
+        }
+
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 15]
```diff
@@ -0,0 +1,635 @@
...
+                # Verify agent name is set correctly
+                assert result[-1].name == agent_name
+
+    @pytest.mark.asyncio
+    async def test_fallback_uses_partial_agent_messages(self):
+        """Test that fallback includes partial agent messages in result."""
+        state = State(messages=[], locale="en-US")
+        current_step = MagicMock()
+        agent_input = {"messages": [HumanMessage(content="Input message")]}
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 16]
```diff
@@ -0,0 +1,635 @@
...
+
+class TestRecursionFallbackEdgeCases:
+    """Test edge cases and boundary conditions for recursion fallback."""
+
+    @pytest.mark.asyncio
+    async def test_fallback_with_empty_observations(self):
+        """Test fallback behavior when there are no observations."""
+        state = State(messages=[], locale="en-US")
+        current_step = MagicMock()
+        agent_input = {"messages": []}
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 17]
```diff
@@ -0,0 +1,635 @@
...
+
+            # With empty messages, should return empty list
+            assert result == []
+
+    @pytest.mark.asyncio
+    async def test_fallback_with_very_long_recursion_limit(self):
+        """Test fallback with very large recursion limit value."""
+        state = State(messages=[], locale="en-US")
+        current_step = MagicMock()
+        agent_input = {"messages": []}
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 18]
```diff
@@ -0,0 +1,635 @@
...
+            with patch("src.graph.nodes.get_llm_by_type") as mock_get_llm, \
+                 patch("src.graph.nodes.get_system_prompt_template") as mock_get_system_prompt, \
+                 patch("src.graph.nodes.sanitize_tool_response", return_value=mock_llm_response.content):
+
+                mock_llm = MagicMock()
+                mock_llm.invoke = MagicMock(return_value=mock_llm_response)
+                mock_get_llm.return_value = mock_llm
+                mock_get_system_prompt.return_value = "Template"
+
+                result = await _handle_recursion_limit_fallback(
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 19]
```diff
@@ -0,0 +1,635 @@
...
+             patch("src.graph.nodes.get_system_prompt_template") as mock_get_system_prompt, \
+             patch("src.graph.nodes.sanitize_tool_response", return_value=mock_llm_response.content):
+
+            mock_llm = MagicMock()
+            mock_llm.invoke = MagicMock(return_value=mock_llm_response)
+            mock_get_llm.return_value = mock_llm
+            mock_get_system_prompt.return_value = "Template"
+
+            # Should not raise, should use default locale
+            result = await _handle_recursion_limit_fallback(
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 20]
```diff
@@ -0,0 +1,635 @@
...
+
+"""
+Unit tests for recursion limit fallback functionality in graph nodes.
+
+Tests the graceful fallback behavior when agents hit the recursion limit,
+including the _handle_recursion_limit_fallback function and the
+enable_recursion_fallback configuration option.
+"""
+
+from unittest.mock import AsyncMock, MagicMock, patch
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 21]
```diff
@@ -0,0 +1,635 @@
...
+
+import pytest
+from langchain_core.messages import AIMessage, HumanMessage
+from langgraph.errors import GraphRecursionError
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: tests/unit/graph/test_nodes_recursion_limit.py [hunk 22]
```diff
@@ -0,0 +1,635 @@
...
+from langgraph.types import Command
```
[the remark this answers is not in the document]
Comment (xunliu) [reply tests/unit/graph/test_nodes_recursion_limit.py]: fixed

File: src/config/configuration.py [hunk 23]
```diff
@@ -63,6 +63,9 @@ class Configuration:
     interrupt_before_tools: list[str] = field(
         default_factory=list
     )  # List of tool names to interrupt before execution
+    enable_recursion_fallback: bool = (
```
Comment (WillemJiang) [inline src/config/configuration.py:66]: Need to update the docs/configuration_guide.md for the newly introduced parameter

**Comment by xunliu** [comment]
> @xunliu, Do you mind updating the document for the newly added configuration?

OK

File: src/config/configuration.py (hunk 23 above)
Comment (xunliu) [reply src/config/configuration.py:66]: fixed

**Comment by WillemJiang** [comment]
Thanks for the contribution of @lxl0413 and @xunliu
